hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/scheduler/ContainerScheduler.java

```diff
@@ -0,0 +1,393 @@
+          continue;
+        }
+        extraOpportContainersToKill.add(runningCont);
+        ResourceUtilizationManager.decreaseResourceUtilization(
+            getContainersMonitor(), resourcesToFreeUp,
+            runningCont.getResource());
+      }
+    }
+    if (!hasSufficientResources(resourcesToFreeUp)) {
+      LOG.warn("There are no sufficient resources to start guaranteed [{}]" +
+          "even after attempting to kill all running" +
+          "opportunistic containers.", containerToStartId);
+    }
+    return extraOpportContainersToKill;
+  }
+
+  private boolean hasSufficientResources(
+      ResourceUtilization resourcesToFreeUp) {
+    return resourcesToFreeUp.getPhysicalMemory() <= 0 &&
+        resourcesToFreeUp.getVirtualMemory() <= 0 &&
+        resourcesToFreeUp.getCPU() <= 0.0f;
+  }
+
+  private ResourceUtilization resourcesToFreeUp(
+      ContainerId containerToStartId) {
+    // Get allocation of currently allocated containers.
+    ResourceUtilization resourceAllocationToFreeUp = ResourceUtilization
+        .newInstance(this.utilizationManager.getCurrentUtilization());
+
+    // Add to the allocation the allocation of the pending guaranteed
+    // containers that will start before the current container will be started.
+    for (Container container : queuedGuaranteedContainers.values()) {
+      ResourceUtilizationManager.increaseResourceUtilization(
+          getContainersMonitor(), resourceAllocationToFreeUp,
+          container.getResource());
+      if (container.getContainerId().equals(containerToStartId)) {
+        break;
+      }
+    }
+
+    // These Resources have already been freed, due to demand from an
+    // earlier Guaranteed container.
+    for (Container container : oppContainersMarkedForKill.values()) {
+      ResourceUtilizationManager.decreaseResourceUtilization(
+          getContainersMonitor(), resourceAllocationToFreeUp,
+          container.getResource());
+    }
+
+    // Subtract the overall node resources.
+    getContainersMonitor().subtractNodeResourcesFromResourceUtilization(
+        resourceAllocationToFreeUp);
+    return resourceAllocationToFreeUp;
+  }
+
+  public void updateQueuingLimit(ContainerQueuingLimit limit) {
+    this.queuingLimit.setMaxQueueLength(limit.getMaxQueueLength());
+    // TODO: Include wait time as well once it is implemented
+    if (this.queuingLimit.getMaxQueueLength() > -1) {
+      shedQueuedOpportunisticContainers();
+    }
+  }
+
+  private void shedQueuedOpportunisticContainers() {
+    int numAllowed = this.queuingLimit.getMaxQueueLength();
+    Iterator<Container> containerIter =
+        queuedOpportunisticContainers.values().iterator();
+    while (containerIter.hasNext()) {
+      Container container = containerIter.next();
+      if (numAllowed <= 0) {
+        container.sendKillEvent(
+            ContainerExitStatus.KILLED_BY_CONTAINER_SCHEDULER,
+            "Container Killed to make room for Guaranteed Container.");
```

Comment:
The Log.info line is printed immediately, which is why it has 'will be killed'.
The diagnostic line is part of the Container status message and is in past-tense and does not need the containerid.